roscala/src/main/scala/coop/rchain/rosette/Meta.scala

```diff
@@ -1,27 +1,75 @@
 package coop.rchain.rosette
 
-import cats.data.State._
-import coop.rchain.rosette.Ctxt.CtxtTransition
+import cats.data.StateT
+import cats.implicits._
+import cats.{Applicative, MonadError}
+import coop.rchain.rosette.Location.valWrt
+import coop.rchain.rosette.utils.implicits.LiftOps
 
 object Meta {
+  def inspect[F[_]: Applicative, A](f: VMState => A) = StateT.inspect[F, VMState, A](f)
+  def pure[F[_]: Applicative](ob: Ob)                = StateT.pure[F, VMState, Ob](ob)
+
   case class StdMeta(meta: Ob, parent: Ob, override val extension: StdExtension) extends Actor {
-    // TODO:
-    def get(client: Ob, key: Ob): CtxtTransition[Result] =
-      pure[Ctxt, Result](Left(Absent))
+    lazy val map = extension.slot.head.asInstanceOf[RblTable]
 
-    // TODO:
-    def lookupOBOStdMeta(client: Ob, key: Ob): CtxtTransition[Result] =
-      for {
-        result <- get(client, key)
-      } yield {
-        result.left.flatMap {
-          // TODO:
-          // BASE(BASE(client)->parent())->lookup(key, ctxt)
-          case Absent if client.isInstanceOf[RblFloat] =>
-            Right(prim.Prims(197))
-          case Absent => Right(prim.Prims(226))
-        }
+    def keyLoc(key: Ob, client: Ob): Location = {
+      val atom = map.getKey(key)
+
+      if (atom == Ob.ABSENT) {
+        Limbo
+      } else {
+        atom.asInstanceOf[Location]
+      }
+    }
+
+    def get[F[_]](client: Ob, key: Ob)(
+        implicit E: MonadError[F, RblError]
+    ): StateT[F, VMState, Ob] = {
+      val container = client
+
+      def extension(ob: Ob, indirect: Boolean): Option[Ob] =
+        Some(ob)
+          .filter(_ => indirect)
+          .map(_.asInstanceOf[Actor].extension)
+
+      def matchLocation(globalEnv: TblObject): Location => StateT[F, VMState, Ob] = {
+        case LexVariable(indirect, _, offset) =>
+          pure[F](
+            extension(container, indirect)
+              .getOrElse(container)
+              .slot(offset))
+        case Limbo =>
+          Absent.liftE[F, RblError]
+        case location =>
+          StateT.liftF[F, VMState, Ob](valWrt[F](location, client).run(globalEnv))
       }
+
+      val loc = keyLoc(key, client)
+
+      for {
+        globalEnv <- inspect[F, TblObject](_.globalEnv)
+        res       <- matchLocation(globalEnv)(loc)
+      } yield res
+    }
+
+    def lookupOBOStdMeta[F[_]](client: Ob, key: Ob)(
+        implicit E: MonadError[F, RblError]): StateT[F, VMState, Ob] = {
```

Comment:
Let's try to keep computation as "local" as possible. Can we make this `StateT[F, Ctxt, Ob]` or `StateT[F, (Ctxt, GlobalEnv), Ob]`. It helps with reasoning and maybe also in the case of multi threading.